{
 "metadata": {
  "name": "",
  "signature": "sha256:7f202854662593db8e426c3ff3d5697d85edb457f49ebae2915b5ee91751dd0f"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "**Computational Biology: Genomes, Networks, Evolution** [MIT OpenCourseware](http://ocw.mit.edu/courses/electrical-engineering-and-computer-science/6-047-computational-biology-genomes-networks-evolution-fall-2008/index.htm)\n",
      "\n",
      "1. Introduction: biology, algorithms, machine learning\n",
      "2. Global/local alignment, dynamic programming\n",
      "3. String search, BLAST, database search\n",
      "4. Clustering basics, gene expression, sequence clustering\n",
      "5. Classification, feature selection, SVM\n",
      "6. HMMs 1: evaluation parsing (hidden markov)\n",
      "7. HMMs 2: posterior decoding, learning\n",
      "8. Recitation: posterior decoding review, Baum-Welch learning\n",
      "9. Generalized HMMs and gene prediction\n",
      "10. Regulatory motifs, Gibbs sampling, EM\n",
      "11. Recitation: entropy, information, background models\n",
      "12. Gene evolution: phylogenetic algorithms, NJ, ML, parsimony\n",
      "13. Molecular evolution, coalescence, selection, Ka/Ks\n",
      "14. Recitation: gene trees, species trees, reconciliation\n",
      "15. Population genomics: fundamentals\n",
      "16. Population genomics: association studies\n",
      "17. Recitation: population genomics\n",
      "18. Genome assembly, Euler graphs\n",
      "19. Comparitive genomics 1: biological signal discovery, evolutionary signatures\n",
      "20. Comparitive genomics 2: phylogenetics, gene and genome duplication\n",
      "21. Conditional random fields, gene finding, feature finding\n",
      "22. Regulatory networks, Bayesian networks\n",
      "23. Inferring biological networks, graph isomorphism, network motifs\n",
      "24. Metabolic modeling 1: dynamic systems modeling\n",
      "25. Metabolic modeling 2: flux balance analysis and metabolic control analysis\n",
      "26. Systems biology\n",
      "27. Synthetic biology\n",
      "\n",
      "- Texts:\n",
      "    - Durbin, et al. (1999) *Biological Sequence Analysis*\n",
      "    - Jones, Pavel, (2004) *An Introduction to Bioinformatics Algorithms*\n",
      "    - Duda, Hart, Stork (2000) *Pattern Classification*\n",
      "\n"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "**Foundations of Computational and Systems Biology** [MIT OpenCourseware](http://ocw.mit.edu/courses/biology/7-91j-foundations-of-computational-and-systems-biology-spring-2014/index.htm)\n",
      "\n",
      "1. Course Introduction; Overview\n",
      "2. Local Alignment; Statistics\n",
      "3. Global Alignment of Protein Statistics\n",
      "4. Comparative Genomics\n",
      "5. Read Alignment\n",
      "6. Genome Assembly\n",
      "7. ChIP-seq/IDR\n",
      "8. RNA-seq Analysis\n",
      "9. Modeling and Discovery of Sequence Motifs (Gibbs Sampler, Alternatives)\n",
      "10. Markov and Hidden Markov Models\n",
      "11. RNA Secondary Structure Prediction\n",
      "12. Introduction to Protein Structure\n",
      "13. Predicting Protein Structure\n",
      "14. Predicting Interactions\n",
      "15. Gene Regulatory Networks\n",
      "16. Protein Interaction Networks\n",
      "17. Logic Modeling of Cell Signaling Networks\n",
      "18. Analysis of Chromatin Structure\n",
      "19. Discovering Quantitative Trait Loci (QTLs)\n",
      "20. Genome Wide Associate Studies\n",
      "21. Synthetic Biology: From Parts to Modules to Therapeutic Systems\n",
      "22. Causality, Natural Computing, and Engineering Genomes\n",
      "\n",
      "- Readings: http://ocw.mit.edu/courses/biology/7-91j-foundations-of-computational-and-systems-biology-spring-2014/readings/\n",
      "\n"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "**Foundations of Algorithms and Computational Techniques in Systems Biology** [MIT OpenCourseware](http://ocw.mit.edu/courses/biological-engineering/20-482j-foundations-of-algorithms-and-computational-techniques-in-systems-biology-spring-2006/index.htm)\n",
      "\n",
      "- Protein Modeling\n",
      "    1. Introduction and Protein Modeling Questions\n",
      "    2. Models of Proteins\n",
      "    3. Discrete Conformational Search\n",
      "    4. Binding and Docking\n",
      "    5. Molecular Dynamics Simulation\n",
      "    6. Molecular Dynamics and Electrostatics\n",
      "    7. Continuum Electrostatic Modeling I\n",
      "    8. Continuum Electrostatic Modeling II\n",
      "    9. Electrostatic Optimization and Design\n",
      "    10. Electrostatics Modeling\n",
      "    11. Statistical Mechanics\n",
      "- Network Modeling\n",
      "    1. Statistical mechanics (cont)\n",
      "    2. Formulating and Simulating Network Models in Biology\n",
      "    3. Formulating Models\n",
      "    4. Nonlinear Dynamics and Stability\n",
      "    5. Steady-State Problems\n",
      "    6. Parameter Fitting and Estimation\n",
      "    7. Paramter Estimation; Robustness, Fragility, Control\n",
      "- Image Modeling\n",
      "    1. 2-D and 3-D Light Microscopy; Image Reconstruction\n",
      "    2. Deconvolution\n",
      "    3. Deconvolution II\n",
      "    4. Blind Deconvolution\n",
      "    5. Optical Flow\n",
      "    6. High-Throughput Data and its Analysis\n",
      "    7. Inference and Statistics\n",
      "    "
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "- **Foundations of Computational Biology** [page](http://ocw.mit.edu/courses/biology/7-91j-foundations-of-computational-and-systems-biology-spring-2014/)\n",
      "    - **Genoimic Analysis**\n",
      "        1. Course Introduction: History of Computational Biology\n",
      "        2. Local Alignment (BLAST) and Statistics\n",
      "        3. Statistics; Significance Testing; Bonferroni Correction\n",
      "        4. Global Alignment of Protein Sequences\n",
      "        5. Comparative Geomic Analysis of Gene Regulation\n",
      "    - **Genomic Analysis - Next Generation Sequencing**\n",
      "        1. Library Complexity and Short Read Alignment\n",
      "        2. Burrows-Wheeler Transform and Alignments\n",
      "        3. Genome Assembly\n",
      "        4. ChIP-seq Analysis; DNA-protein Interactions\n",
      "    - **Modeling Biological Functioning**\n",
      "        1. Modeling and Discovery of Sequence Motifs (Gibbs Sampler, Alternatives)\n",
      "        2. Gene Expression Program Discovery Using Topic models\n",
      "        3. Markov and Hidden Markov Models of Genomic and Protein Features\n",
      "        4. RNA Secondary Structure - Biological Functions and Prediction\n",
      "        5. Probabilistic Grammatical Models of RNA Structure\n",
      "    - **Proteomics**\n",
      "        1. Introduction to Protein Structure; Structure Comparison and Classification\n",
      "        2. Protein Amino Acid Sidechain Packing Using Markoc random Fields\n",
      "        3. Predicting Protein Structure\n",
      "        4. Predicting Protein Interactions\n",
      "        5. Protein/Protein Interaction Prediction Using Threading\n",
      "    - **Regulatory Networks**\n",
      "        1. Gene Regulatory Networks\n",
      "        2. Protein Interaction Networks\n",
      "        3. Regression Trees\n",
      "        4. Logic Modeling of Cell Signaling Networks\n",
      "        5. Analysis of Chromatin Structure\n",
      "        6. BayesNets\n",
      "    - **Computational Genetics**\n",
      "        1. Discovering Quantitative Trait Loci (QTLs)\n",
      "        2. Narrow Sense Heritability\n",
      "        3. Human genetics, SNPs, and Genome Wide Associate Studies\n",
      "        4. Synthetic Biology: From Parts to Modules to Therapeutic Systems\n",
      "        5. Causality, Natural Computing, and Engineering Genomes\n",
      "        \n",
      "- **Systems Biology** [page](http://ocw.mit.edu/courses/physics/8-591j-systems-biology-fall-2004/)\n",
      "    - **I. Systems Microbiology - 'The Cell as a Well-Stirred Bioreactor'**\n",
      "        1. Introduction; Michaelis-Mentes Kinetics [page](http://ocw.mit.edu/courses/physics/8-591j-systems-biology-fall-2004/readings/l2_syllabus.pdf) [matlab code](http://ocw.mit.edu/courses/physics/8-591j-systems-biology-fall-2004/readings/CodeI1_v2.m)\n",
      "        2. Equilibrium Binding; Cooperativity\n",
      "        3. Lambda Phage; Multistability [page](http://ocw.mit.edu/courses/physics/8-591j-systems-biology-fall-2004/readings/l3_syllabus.pdf) [matlab code](http://ocw.mit.edu/courses/physics/8-591j-systems-biology-fall-2004/readings/CodeI2.m)\n",
      "            - Hasty, Jeff, Joel Pradines, Milos Dolnik, and J. J. Collins. \"Noise-based Switches and Amplifiers for Gene Expression.\" Proc. Natl. Acad. Sci. USA 97, no. 5 (Feb 29, 2000): 2075-80.\n",
      "            - Isaacs, Farren J., Jeff Hasty, Charles R. Cantor, and J. J. Collins. \"Prediction and Measurement of an Autoregulatory Genetic Module.\" PNAS 100, no. 13 (June 24, 2003): 7714-19.\n",
      "        4. Synthetic Genetic Switches [page](http://ocw.mit.edu/courses/physics/8-591j-systems-biology-fall-2004/readings/l5_syllabus.pdf)\n",
      "            - Gardner, Timothy S., Charles R. Cantor, and James J. Collins. \"Construction of a Genetic Toggle Switch in Escherichia coli.\" Nature 403, no. 6767 (January 20, 2000): 339-42.\n",
      "        5. Stability Analysis [page](http://ocw.mit.edu/courses/physics/8-591j-systems-biology-fall-2004/readings/l6_syllabus.pdf)\n",
      "        6. Introduction E. Coli Chemotaxis [page](http://ocw.mit.edu/courses/physics/8-591j-systems-biology-fall-2004/readings/l8_syllabus.pdf) [matlab code](http://ocw.mit.edu/courses/physics/8-591j-systems-biology-fall-2004/readings/CodeI4.m)\n",
      "        7. Fine-tuned versus Robust Chemotaxis\n",
      "            - Spiro, Peter A., John S. Parkinson, and Hans G. Othmer. \"A Model of Excitation and Adaptation in Bacterial Chemotaxis.\" Proc. Natl. Acad. Sci. USA 94, no. 14 (July, 1997): 7263\u201368.\n",
      "        8. Wrapping up Chemotaxis\n",
      "            - Biological Oscillators [page](http://ocw.mit.edu/courses/physics/8-591j-systems-biology-fall-2004/readings/l9_syllabus.pdf) [matlab code](http://ocw.mit.edu/courses/physics/8-591j-systems-biology-fall-2004/readings/CodeI6.m)\n",
      "        9. Genetic Oscillators\n",
      "            - Routh-Hurwitz [page](http://ocw.mit.edu/courses/physics/8-591j-systems-biology-fall-2004/readings/l11_syllabus.pdf) [matlab code](http://ocw.mit.edu/courses/physics/8-591j-systems-biology-fall-2004/readings/routh.m)\n",
      "            - Elowitz, Michael B., and Stanislas Leibler. \"A Synthetic Oscillatory Network of Transcriptional Regulators.\" Nature 403, no. 6767 (January 20, 2000): 335-8.\n",
      "        10. Stochastic Chemical Kinetics\n",
      "            - Atkinson, Mariette R., Michael A. Savageau, Jesse T. Myers, and Alexander J. Ninfa. \"Development of Genetic Circuitry Exhibiting Toggle Switch or Oscillatory Behavior in Escherichia coli.\" Cell 113, no. 5 (May 30, 2003): 597-607.\n",
      "            - Oudenaarden, 2004, The origin and cosequences of noise in biochemical systems [page](http://ocw.mit.edu/courses/physics/8-591j-systems-biology-fall-2004/readings/l12_syllabus.pdf)\n",
      "    - **II. Cell Systems Biology - 'The Importance of Diffusion and Gradients for Cellular Regulation**\n",
      "        1. Introduction to Cell Systems Biology; Fick's Laws\n",
      "        2. Local Excitation; Global Inhibition Theory [page](http://ocw.mit.edu/courses/physics/8-591j-systems-biology-fall-2004/readings/l15_syllabus.pdf)\n",
      "        3. Rapid Pole-to-pole Oscillations in E. Coli\n",
      "        4. Models for Eukaryotic Gradient Sensing [page](http://ocw.mit.edu/courses/physics/8-591j-systems-biology-fall-2004/readings/l19_syllabus.pdf) [matlab code](http://ocw.mit.edu/courses/physics/8-591j-systems-biology-fall-2004/readings/narang.m)\n",
      "        5. Modeling Cytoskeleton Dynamics\n",
      "    - **III. Developmental Systems Biology - 'Building an Organism Starting from a Single Cell'**\n",
      "        1. Quarum Sensing\n",
      "            - Miller, Bassler, 2001, Quorum Sensing in Bacteria [paper](http://everettsd.schoolwires.net/cms/lib07/WA01920133/Centricity/Domain/1010/Quorum%20Sensing%20-%20Bassler%20Article.pdf)\n",
      "            - Andrianantoandro, et al. 2006, Synthetic biology:new engineering rules for an emerging discipline [paper](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.281.89&rep=rep1&type=pdf)\n",
      "        2. Final Problem Set Question Hour\n",
      "        3. Drosophila Development"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "- **Foundations of Algorithms and Computational Techniques in Systems Biology** [page](http://ocw.mit.edu/courses/biological-engineering/20-482j-foundations-of-algorithms-and-computational-techniques-in-systems-biology-spring-2006/)\n",
      "    - **Protein Modeling**\n",
      "        1. Introduction to Protein Modeling Questions\n",
      "        2. Modeling of Proteins\n",
      "        3. Discrete Conformational Search\n",
      "        4. Binding and Docking\n",
      "        5. Molecular Dynamics Simulation\n",
      "        6. Molecular Dynamics and Electrostatics\n",
      "        7. Continuum Electrostatic Modeling I\n",
      "        8. Continuum Electrostatic Modeling II\n",
      "        9. Electrostatic Optimization and Design\n",
      "        10. Electrostatics Modeling\n",
      "        11. Statistical Mechanics\n",
      "    - **Network Modeling**\n",
      "        1. Formulating and Simulating Network Models in Biology\n",
      "        2. Formulating Models\n",
      "        3. Nonlinear Dynamics and Stability\n",
      "        4. Steady-State Problems\n",
      "        5. Parameter Fitting and Estimation\n",
      "        6. Parameter Estimation; Robustness, Fragility, Control\n",
      "    - **Image Modeling**\n",
      "        1. 2D and 3D Light Microscopy; Image Reconstruction\n",
      "        2. Deconvolution\n",
      "        3. Deconvolution II\n",
      "        4. Blind Deconvolution\n",
      "        5. Optical Flow\n",
      "        6. High Throughput Data and its Analysis\n",
      "        7. Inference and Statistics\n",
      "        8. Roundup\n",
      "        "
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "- **Design for a Darwinian Brain** - Chrisantha Fernando [YouTube talk](https://www.youtube.com/watch?v=L9SqrLTf0JA) [Google Scholar](https://scholar.google.co.uk/citations?user=02f6uacAAAAJ&hl=en)\n",
      "    - *A New Research Program: Evolutionary Neurodynamics* [page](http://simons.berkeley.edu/talks/chrisantha-fernando-2014-03-19)\n",
      "    - *What is comparable in comparative cognition* [page](http://chittkalab.sbcs.qmul.ac.uk/2012/Chittka%20et%20al%202012%20PhilTrans.pdf)\n",
      "    - *Evolution of associative learning in chemical networks* [page](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002739)\n",
      "    - *Is there a liquid state machine in the bacterium E. Coli?* [page](http://www.cs.bham.ac.uk/~jer/papers/04218885.pdf)\n",
      "    - *Evolvable physical self-replicators* [page](http://www.mitpressjournals.org/doi/abs/10.1162/artl_a_00056#.VPicdimrl8M)\n",
      "    - *The Theory of Darwinian Neurodynamics* [page](https://www.youtube.com/watch?v=J0c7g491PCE)\n",
      "    - *Cognitive Robotics* [page](https://www.youtube.com/watch?v=CU251tm7UI8)\n",
      "    - Talk given at the 2013 Intelligent Sensing Summer School [page](http://cis.eecs.qmul.ac.uk/school2013.html)\n",
      "    - *The neuronal replicator hypothesis*\n",
      "        - There are replicators in your brain\n",
      "        - They evolve by natural selection as you think and sleep\n",
      "        - They allow open-ended thought by generation and selection of certain idea/action and value representations in your brain\n",
      "        - We think the above is true because of a convergence of implementation and algorithmic arguments\n",
      "        - We are now testing the hypothesis on several fronts:\n",
      "            - Neuronal (Model and find replicators in the brain)\n",
      "            - Cognitive (Fit human behavior with the above models)\n",
      "            - Constructive (Make something that uses the above algorithms that does things better than anyone else)\n",
      "    - *What is natural selection?*\n",
      "        - John Maynard Smith - units of evolution multiply - > *but this is not sufficient to describe evolultion, since there is no selection. Fire multiplies but is not selected upon*\n",
      "            - you also need heredity (offspring resemble parents more than others) \n",
      "            - you also need *variation* (differences within a generation - variability is differnces over multiple generations)\n",
      "            - **Therefore**, units of evolution have multiplication, variation and heredity, with selection yielding higher fitness\n",
      "        - *We need to distinguish natural selection from many other algorithms that **are not** natural selection*\n",
      "        - Solitary search - (stochastic) hill-climbing\n",
      "        - Parallel search - Independent hill climbers (typically a linear speed up)\n",
      "        - Parallel search with Competition (price) \n",
      "            1. Competitive learning (a robot/agent only has a certain lifespan/can only take a certain number of steps)\n",
      "                - this is where edelmann's neural darwinism could be categorized\n",
      "            2. Reinforcement Learning\n",
      "            3. Synaptic Selectionism\n",
      "            4. Neural Darwinism\n",
      "        - Parallel search with Competition and Information Transmission (between agents)\n",
      "            1. Genetric Natural Selection\n",
      "            2. Adaptive Immune System\n",
      "            3. Genetic Algorithms\n",
      "            4. Didactic Receptive fields\n",
      "            5. Natural Replicators\n",
      "        - *Selectionist and Evolutionary approaches to brain function: a critical appraisal* [page](http://journal.frontiersin.org/article/10.3389/fncom.2012.00024/abstract)\n",
      "    - *Computational Modeling*\n",
      "        - Copying and Evolution of Neuronal Topology [page](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.278.1395&rep=rep1&type=pdf)\n",
      "            - \"small patterns of connectivity can copy themselves from one part of the brain to another\"\n",
      "            - need to worry about \"Eigen's error catastrophe\", which is the idea that the accumulation of successive mutations could lead to the complete malfunction of an organism and its subsequent inability to reproduce (extinction). [wiki](http://en.wikipedia.org/wiki/Error_catastrophe) The phenomena is predicted mathematically and observed empirically. \n",
      "                - \"if it makes too many mutations, it may lose some of its biological features which have evolved to its advantage, including its ability to reproduce at all. The question arises: how many mutations can be made during each replication before the population of viruses begins to lose self-identity?\"\n",
      "        - *Evolvable Neuronal Paths: A Novel Basis for Information and Search in the Brain* [page](http://chittkalab.sbcs.qmul.ac.uk/VeraPub/Fernando%20et%20al%202011%20PLoS%20ONE.pdf)\n",
      "            - another way to get copying: if you have a network, there are many possible pathways of activation through that networks.\n",
      "            - if every path is a genotype - multiple different inheritable characteristics of information processing \"embodied\" in a network.\n",
      "            - selection on patterns of neural activity    \n",
      "        - *Test 1: model and final neural replicators*\n",
      "        - *Test 2: Model development and generative (creative - high dimensional generative searches) behavior in children*\n",
      "            - we create our own goal functions to solve a creative problem, then we select those ideas. \n",
      "        - *Test 3: Show open-ended creativity in robots*\n",
      "            - create a cognitive architecture that is darwinian and see if it exhibits creative behavior\n",
      "            - subfield = evolutionary robotics (embed a continuous time recurrent neural network in a robot for evolved behavior) \n",
      "                - *Towards Behavior Control for Evolutionary Robot Based RL with ENN* [paper](http://cdn.intechopen.com/pdfs-wm/43602.pdf)\n",
      "      - *A Darwinian Cognitive Architecture* \n",
      "          - aim: black box with different sensory input able to plug into it\n",
      "          - schism between two camps: physical symbol systems (Fogel) and connectionist/neural network people\n",
      "              - Q: what is the minimum structure needed for general intelligence? \n",
      "              - Fogel: you need a physical symbol system - connectionist systems do not have atomic systems with rules of linkages that can be recombined\n",
      "          - Consists of functional 'atoms' capable of replication (a neural microcircuit - piece of cortex)\n",
      "              - atom could be a single classification hierarchy structure from ML\n",
      "              - atom = a neuronal group in cortex with certain functional properties (response characteristics, processing abilities)\n",
      "              - replication of atoms is evident during cortical map plasticity where cortical representations grow and shrink\n",
      "          - These link to form 'molecules' capable of replication\n",
      "              - molecules are functional circuits. They are implemented by cortex (atom) - Basal Ganglia (register/router/edge) - cortex (atom) bindings\n",
      "              - BG could coordinated atoms switching on and off other atoms\n",
      "          - 'Molecules' aggregate into function 'compartments' capable of replication\n",
      "              - these are co-activated groups of molecules that control behavior as a unit. They are an important unit of selection in the brain\n",
      "              - these 'compartments' could be probabilistic, but they are the 'current' modules to which reward is attributed\n",
      "       - **How are compartments selected?**\n",
      "           - Babies reset themselves to an initial condition between 'trials'. They often repeat a trial with the same goal many times with variation\n",
      "           - The goal may be hard to understand to an external observer (difficult to guess their goal function)\n",
      "           - Given a goal/task, we abstract the child's action selection mechanism as a microbial genetic alrogithm\n",
      "               - we can abstract their search\n",
      "               - pick to compartments at random, and judge the movement they allow the robot to control for a time wrt a fitness function\n",
      "       - **A first attempt at evolving compartments**\n",
      "           - **Evolutionary Robot Operating System (EROS)**\n",
      "               - Python framework for evolving robot controllers\n",
      "               - Works with Nao, e-puck (webots), Arduino robots\n",
      "               - Aim: plug it into any robot, and it works out interesting things to do with it, like a baby brain with a baby body (more on this later)\n",
      "               1. initialize a opulation of actor compartments\n",
      "                   - izhikevich neural networks\n",
      "               2. Evolve the population\n",
      "                   - Operators used:\n",
      "                       A. Sensor Atom\n",
      "                           - add/remove sensor from sensor list\n",
      "                       B. Linear Transform Atom\n",
      "                           - Gaussian mutation to elements with at 5% per element\n",
      "                       C. Izhikevich Transform Atom\n",
      "                           - Change input/output groups\n",
      "                           - Gaussian mutation to linear readouts\n",
      "                       D. Oscillator Atom\n",
      "                           - Change frequency/amplitude of oscillation\n",
      "                           - Add/remove oscillators\n",
      "                   - Add S-LT-M, S-Izh-M, S-O-LT-M, molecule 5%\n",
      "                   - Remove random group 1% x molecule number in compartment\n",
      "                   - Add/Remove random edge\n",
      "      - **The Goal of Life-time Learning**\n",
      "          - can we use these evolved molecules on new tasks?\n",
      "          - Physical Symbol Systems, Modularity, Reuse, Compositionality, Systematicity\n",
      "              - Q: We've got molecules in compartments but do they really behave like molecules, i.e. do they exhibit systematicity and compositionality?\n",
      "              - A: Store the molecules from the best evolved compartment after and evolutionary run for Max Z, in an archive. During all subsequent evolution add complete molecules from the acheive to the mutated winning compartment with 5% probability\n",
      "      - **Evolving Games/Fitness Functions**\n",
      "          - The algorithm didn't invent the game to play. We imposed it.\n",
      "          - Can an algorithm decide which games to play?\n",
      "          - How should it decide what a good game to play is?\n",
      "          - As academics we also invent our research problems and try to solve them\n",
      "          - Clever children like clever academics choose the right games to play\n",
      "          - Q: How do you choose the games to play?\n",
      "          - What is a good game/goal for a baby?\n",
      "              - Not too hard\n",
      "                  - all solutions have low fitness on the game (floor effect)\n",
      "              - Not to easy\n",
      "                  - all solutions have high fitness on the game (ceiling effect)\n",
      "              - just right **(progress is being made, and further progress is predicted)**\n",
      "                  - maximum variance in the quality of modified (best guessed) novel solutions\n",
      "                  - This is deeply related to **Fisher's fundamental theorem of natural selection: The rate of fitness increase is proportional to the fitness variance** (the more differnce there is in the fitnes of a population the greater rate of fitness change)\n",
      "                  - we can evolve games with this variance as a metric\n",
      "                  - but it has some stipulations that dont actually occur in real systems (since most evolutionary change is neutral)\n",
      "                  - Given some assumptions, e.g. no linkage disequilibrium, no frequency-dependent selection, no epistasis (yeah right)\n",
      "                  - neutral exploration is more common, resulting in **punk eak**, and making evolutionary progress hard to predict.\n",
      "       - **General Minimal Criteria for Good Games**\n",
      "           - Absolute criteria:\n",
      "               - A game is not worth playing unless progress is being made on it, or if progress is not predicts [Progress criterion, Oudayer]\n",
      "               - A game is not worth playing if it likely to kill you, cause you irreparable damage [Variability criterion, Ashby]\n",
      "           - Relative criteria:\n",
      "               - a good game allows learning of information useful for real challenges later [Latent learning criterion, Tolman]\n",
      "               - A good game builds knowledge about the world: allows facts to be learned [Epistemic criterion, Barto]\n",
      "               - A good game builds a deep understanding about the world, allows generative models (theories) to be learned [Compression criterion, Schmidhuber - interest is the first derivative of compression]\n",
      "               - A good game allows one to predict more about the future i.e. maximizes that first derivative of predictive information [Predictive information criterion, Der, Mathius] (how much uncertainty about the future can you minimize given the current state + you know the past)\n",
      "               - A good game allows many novel solutions to be produced quickly [Novelty criterion, Lehman and Stanley] (removed the fitness function, only reward an individual if they produce distant behaviors from other solutions according to some measure) \n",
      "               - ** Current problem: open-ended evolution for continual novelty**\n",
      "               \n",
      "        - What's missing\n",
      "            - imitation and social learning"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}